### PR TITLE
Issues/3916 export all slurm

### PR DESCRIPTION
**Pull request by Hexotical**
## Changelog Entry
To be copied to the [draft changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog) by merger:

Make --export-all the default for slurm arguments

## Reviewer Checklist

 * [ ] Make sure it is coming from `issues/XXXX-fix-the-thing` in the Toil repo, or from an external repo.
    * [ ] If it is coming from an external repo, make sure to pull it in for CI with:
        ```
        contrib/admin/test-pr otheruser theirbranchname issues/XXXX-fix-the-thing
        ```
    * [ ] If there is no associated issue, [create one](https://github.com/DataBiosphere/toil/issues/new).
* [ ] Read through the code changes. Make sure that it doesn't have:
    * [ ] Addition of trailing whitespace.
    * [ ] New variable or member names in `camelCase` that want to be in `snake_case`.
    * [ ] New functions without [type hints](https://docs.python.org/3/library/typing.html).
    * [ ] New functions or classes without informative docstrings.
    * [ ] Changes to semantics not reflected in the relevant docstrings.
    * [ ] New or changed command line options for Toil workflows that are not reflected in `docs/running/{cliOptions,cwl,wdl}.rst` 
    * [ ] New features without tests.
* [ ] Comment on the lines of code where problems exist with a review comment. You can shift-click the line numbers in the diff to select multiple lines.
* [ ] Finish the review with an overall description of your opinion.

## Merger Checklist

* [ ] Make sure the PR passes tests.
* [ ] Make sure the PR has been reviewed **since its last modification**. If not, review it.
* [ ] Merge with the Github "Squash and merge" feature.
    * [ ] If there are multiple authors' commits, add [Co-authored-by](https://github.blog/2018-01-29-commit-together-with-co-authors/) to give credit to all contributing authors.
* [ ] Copy its recommended changelog entry to the [Draft Changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog).
* [ ] Append the issue number in parentheses to the changelog entry.

### CODE REVIEW

**Comment by adamnovak** [comment]
This would fix #3916.

File: docs/appendices/environment_vars.rst [hunk 1]
```diff
@@ -139,7 +139,7 @@ There are several environment variables that affect the way Toil runs.
 | TOIL_SLURM_ARGS                  | Arguments for sbatch for the slurm batch system.   |
 |                                  | Do not pass CPU or memory specifications here.     |
 |                                  | Instead, define resource requirements for the job. |
-|                                  | There is no default value for this variable.       |
+|                                  | By default, "--export=ALL" is set.                 |
```
Comment (adamnovak) [inline docs/appendices/environment_vars.rst:142]: This makes it sound like if `TOIL_SLURM_ARGS` is not set, it has the same effect as setting it to `--export=ALL`. In fact, `--export=all` is supposed to be set whether `TOIL_SLURM_ARGS` is set or not. So you should elaborate here and explain how if the `--export` or `--export-file` options are used here, they will override the default behavior of exporting all environment variables.

File: src/toil/batchSystems/slurm.py [hunk 2]
```diff
@@ -302,6 +302,8 @@ def prepareSbatch(self,
                     raise ValueError(f"Some resource arguments are incompatible: {nativeConfig}")
 
                 sbatch_line.extend(nativeConfig.split())
+                if "--export" not in nativeConfig:
```
[the remark this answers is not in the document]
Comment (adamnovak) [reply src/toil/batchSystems/slurm.py:305]: You might also want to change the checks above for `--mem` and `--cpus-per-task` that are just substring checks.

**Review by adamnovak — CHANGES_REQUESTED** [review]
I don't think this will quite work as written. And I think the simple substring check to detect if the option is set already is too easy to trick.

**Comment by adamnovak** [comment]
@Hexotical Do you want to finish this PR?

File: docs/appendices/environment_vars.rst (hunk 1 above)
Comment (adamnovak) [inline docs/appendices/environment_vars.rst:142]:
```suggestion
|                                  | If neither ``--export`` nor ``--export-file`` is   |
|                                  | in the argument list, ``--export=ALL`` will be     |
|                                  | provided.                                          |
```

**Review by adamnovak — APPROVED** [review]
This looks good except for the docs, which I think I have a fix for.

**Comment by adamnovak** [comment]
@thiagogenez We're hoping to merge this as soon as we can get it working, thanks for the help!

Really for this you would want to talk to @Hexotical and not me, though; it looks like we forgot to set the assignee on #3916 but we kind of gave the issue to him to do.

And yeah, `docker exec export` is never going to work, because `exec` starts a new process and `export` is a Bash builtin to set an environment variable for itself. AFAIK only the process itself can set an environment variable on a running process.

**Comment by thiagogenez** [comment]
> @thiagogenez We're hoping to merge this as soon as we can get it working, thanks for the help!
> 
> Really for this you would want to talk to @Hexotical and not me, though; it looks like we forgot to set the assignee on #3916 but we kind of gave the issue to him to do.
> 
> And yeah, `docker exec export` is never going to work, because `exec` starts a new process and `export` is a Bash builtin to set an environment variable for itself. AFAIK only the process itself can set an environment variable on a running process.

Thank you @adamnovak for the message. Glad to help with this. I'm very interested to test Toil on Slurm using Cactus. 

Please @Hexotical count on me if you need a hand for testing etc.

**Pull request merged**